pkg/iss/iss.go

```diff
@@ -419,57 +417,29 @@ func New(
 	})
 
 	chkppbdsl.UponStableCheckpointReceived(iss.m, func(sender t.NodeID, sn tt.SeqNr, snapshot *trantorpbtypes.StateSnapshot, cert map[t.NodeID][]byte) error {
-		_chkp := checkpointpbtypes.StableCheckpoint{
+		_chkp := &checkpointpbtypes.StableCheckpoint{
 			Sn:       sn,
 			Snapshot: snapshot,
 			Cert:     cert,
 		}
-		chkp := checkpoint.StableCheckpointFromPb(_chkp.Pb())
 
-		// Check syntactic validity of the checkpoint.
-		if err := chkp.SyntacticCheck(params); err != nil {
-			iss.logger.Log(logging.LevelWarn, "Ignoring checkpoint. Syntactic check failed.", err)
-			return nil
-		}
+		cvcpbdsl.ValidateCheckpoint(iss.m, iss.moduleConfig.CVC, _chkp, iss.epoch.Nr(), iss.memberships, &validateChechkpointContext{
+			checkpoint: _chkp,
+		})
+		return nil
+	})
 
-		// Ignore checkpoint if we are not part of its membership
-		// (more precisely, membership of the epoch the checkpoint is at the start of).
-		// Correct nodes should never send such checkpoints, but faulty ones could.
-		if _, ok := chkp.Memberships()[0].Nodes[iss.ownID]; !ok {
-			iss.logger.Log(logging.LevelWarn, "Ignoring checkpoint. Not in membership.",
-				"sender", sender, "memberships", chkp.Memberships())
+	cvcpbdsl.UponCheckpointValidated(iss.m, func(err error, c *validateChechkpointContext) error {
+		if err != nil {
+			iss.logger.Log(logging.LevelWarn, "Ignoring checkpoint. Validation failed:", err)
 			return nil
 		}
 
-		// Check how far the received stable checkpoint is ahead of the local node's state.
+		chkp := checkpoint.StableCheckpointFromPb(c.checkpoint.Pb())
```

Comment:
This was probably a sort of brain fart of mine when I wrote the original code. But I now I don't see a reason to convert the checkpoint to PB and back like this...

Reply:
There are three types: DSL-type StableCheckpoint, PB-type StableCheckpoint, and manual go-code with custom functions StableCheckpoint. Here, we are converting from DSL-type to Pb() and giving that to the go-code function that takes Pb as input to convert to go-code type.

Reply:
Aaaah yes I see, my bad.